src/librustc/traits/error_reporting/mod.rs

```diff
@@ -917,14 +917,27 @@ impl<'a, 'tcx> InferCtxt<'a, 'tcx> {
             // already reported in the query
             ConstEvalFailure(err) => {
                 if let ErrorHandled::TooGeneric = err {
-                    // Silence this error, as it can be produced during intermediate steps
-                    // when a constant is not yet able to be evaluated (but will be later).
-                    return;
+                    // In this instance, we have a const expression containing an unevaluated
+                    // generic parameter. We have no idea whether this expression is valid or
+                    // not (e.g. it might result in an error), but we don't want to just assume
+                    // that it's okay, because that might result in post-monomorphisation time
+                    // errors. The onus is really on the caller to provide values that it can
+                    // prove are well-formed.
+                    let mut err = self.tcx.sess.struct_span_err(
+                        span,
+                        "cannot determine that constant is always well-formed",
+                    );
+                    // FIXME(const_generics): this is not always possible at the moment.
+                    // When it is, we should give more specific suggestions, like exactly
+                    // what sort of bound needs to be added.
+                    err.note("well-formedness must be enforced by adding it to the bounds");
+                    err.emit();
+                } else {
+                    self.tcx.sess.delay_span_bug(
+                        span,
+                        &format!("constant in type had an ignored error: {:?}", err),
+                    );
                 }
-                self.tcx.sess.delay_span_bug(
-                    span,
-                    &format!("constant in type had an ignored error: {:?}", err),
-                );
                 return;
```

Comment:
Please turn this into an exhaustive match, because then it would make sense (since the other variant explicitly indicates an error has already been reported).